### PR TITLE
Use c++14.

### PR DESCRIPTION
**Pull request by endJunction**
With recent compiler improvements I hope to be able to use the three-year-old standard.

### CODE REVIEW

**Comment by endJunction** [comment]
So, this works on VisualStudio15 too. Of course not all features can be used, but this is still good news.
constexpr, make_unique_ptr, templated lambdas maybe...

**Comment by chleh** [comment]
Could you also add a commit here that uses C++14 features, e.g. using `make_unique` everywhere? Otherwise it's easy for a compiler to claim that it supports C++14. :smile:

**Comment by endJunction** [comment]
@chleh The unique pointer is not available in gcc-4.8.1, so this is based on the new envinf environment.

**Comment by chleh** [comment]
Very nice!

But should we really change `... .reset(new ...)` to `... = make_unique<...>(...)`?

**Comment by endJunction** [comment]
clang-tidy's automatic changes.

File: NumLib/ODESolver/TimeDiscretizationBuilder.cpp [hunk 1]
```diff
@@ -16,36 +16,32 @@ namespace NumLib
...
-        return T(new ConcreteTD);
+        return std::make_unique<ForwardEuler>();
     }
     else if (type == "CrankNicolson")
     {
         //! \ogs_file_param{prj__time_loop__processes__process__time_discretization__CrankNicolson__theta}
         auto const theta = config.getConfigParameter<double>("theta");
-        using ConcreteTD = CrankNicolson;
-        return T(new ConcreteTD(theta));
+        //! \ogs_file_param_special{prj__time_loop__processes__process__time_discretization__CrankNicolson}
```
Comment (chleh) [inline NumLib/ODESolver/TimeDiscretizationBuilder.cpp]: Already documented two lines above as far as i can see.

File: NumLib/ODESolver/TimeDiscretizationBuilder.cpp [hunk 2]
```diff
@@ -16,36 +16,32 @@ namespace NumLib
...
+        return std::make_unique<CrankNicolson>(theta);
     }
     else if (type == "BackwardDifferentiationFormula")
     {
         //! \ogs_file_param{prj__time_loop__processes__process__time_discretization__BackwardDifferentiationFormula__order}
         auto const order = config.getConfigParameter<unsigned>("order");
-        using ConcreteTD = BackwardDifferentiationFormula;
-        return T(new ConcreteTD(order));
+        //! \ogs_file_param_special{prj__time_loop__processes__process__time_discretization__BackwardDifferentiationFormula}
```
Comment (chleh) [inline NumLib/ODESolver/TimeDiscretizationBuilder.cpp]: Likewise. Note: Documentation has to be put above `config.getConfig...(...);` lines. It has to be there and only there. (If I'm not wrong.)

**Comment by wenqing** [comment]
:+1: .
It's OK to increase the vtkdiff tolerance for ParallelFEM_GroundWaterFlow3D_DirichletBC.

File: NumLib/ODESolver/TimeDiscretizationBuilder.cpp (hunk 1 above)
Comment (chleh) [inline NumLib/ODESolver/TimeDiscretizationBuilder.cpp]: Clarifeid in discussion I was wrong. But this probably solves #1816.

**Comment by chleh** [comment]
:+1:

**Comment by endJunction** [comment]
:fireworks: :confetti_ball: :tada:

**Pull request merged**